### PR TITLE
Single variable API for Modify Process Instance command

### LINKED ISSUE #13460: newModifyProcessInstanceCommand: complete command with single variable
**Is your feature request related to a problem? Please describe.**
This feature is related to #13040 . Using Java Client I want to complete `.newModifyProcessInstanceCommand()` by using only a single variable

**Describe the solution you'd like**
`client.newModifyProcessInstanceCommand(job).variable("name", value)`

**Describe alternatives you've considered**
`client.newModifyProcessInstanceCommand(job).variables(Map.of("name", value))`, but `Map.of()` is available only with Java 9 or above

### PR DESCRIPTION
**Pull request by nicpuppa**
## Description

Add a new API to `newModifyInstanceProcess()` command to allow user to provide a single variable. This reduce overhead and improve the user experience

## Related issues

closes #13460 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by nicpuppa** [comment]
Hi @remcowesterhoud, this PR depends on #13442 , I leave it in draft until the other one is merged

File: clients/java/src/main/java/io/camunda/zeebe/client/api/command/ModifyProcessInstanceCommandStep1.java [hunk 1]
```diff
@@ -162,5 +162,31 @@ ModifyProcessInstanceCommandStep3 withVariables(
      * @return the builder for this command
      */
     ModifyProcessInstanceCommandStep3 withVariables(final Object variables, final String scopeId);
+
+    /**
+     * Create a {@link
+     * io.camunda.zeebe.gateway.protocol.GatewayOuterClass.ModifyProcessInstanceRequest.VariableInstruction}
+     * for the element that's getting activated. These variable will be created in the global scope
```
Comment (remcowesterhoud) [inline clients/java/src/main/java/io/camunda/zeebe/client/api/command/ModifyProcessInstanceCommandStep1.java:169]:
```suggestion
     * for the element that's getting activated. This variable will be created in the global scope
```

🔧 small typo

File: clients/java/src/main/java/io/camunda/zeebe/client/api/command/ModifyProcessInstanceCommandStep1.java [hunk 2]
```diff
@@ -162,5 +162,31 @@ ModifyProcessInstanceCommandStep3 withVariables(
...
+     * @param key the key of the variable to be serialized to JSON
+     * @param value the value of the variable to be serialized to JSON
+     * @return the builder for this command
+     */
+    ModifyProcessInstanceCommandStep3 withVariable(final String key, final Object value);
+
+    /**
+     * Create a {@link
+     * io.camunda.zeebe.gateway.protocol.GatewayOuterClass.ModifyProcessInstanceRequest.VariableInstruction}
+     * for the element that's getting activated. These variable will be created in the scope of the
```
Comment (remcowesterhoud) [inline clients/java/src/main/java/io/camunda/zeebe/client/api/command/ModifyProcessInstanceCommandStep1.java:181]:
```suggestion
     * for the element that's getting activated. This variable will be created in the scope of the
```

🔧 Same thing

File: clients/java/src/main/java/io/camunda/zeebe/client/impl/command/ModifyProcessInstanceCommandImpl.java [hunk 3]
```diff
@@ -150,13 +151,34 @@ public ModifyProcessInstanceCommandStep3 withVariables(
...
       final InputStream variables, final String scopeId) {
     ArgumentUtil.ensureNotNull("variables", variables);
     final String variablesString = jsonMapper.validateJson("variables", variables);
     return createVariableInstruction(variablesString, scopeId);
   }
 
+  private VariableInstruction createVariableInstruction(
+      final String key, final Object value, final String scopeId) {
+    ArgumentUtil.ensureNotNull("key", key);
+    ArgumentUtil.ensureNotNull("value", value);
```
Comment (remcowesterhoud) [inline clients/java/src/main/java/io/camunda/zeebe/client/impl/command/ModifyProcessInstanceCommandImpl.java:177]: ❌ Is there a reason why you decided to not allow `null` as a value?

Comment (remcowesterhoud) [reply clients/java/src/main/java/io/camunda/zeebe/client/impl/command/ModifyProcessInstanceCommandImpl.java:177]: I just realised I missed this in the initial PR as well. `null` is a valid value so please remove it there as well.

File: clients/java/src/main/java/io/camunda/zeebe/client/impl/command/ModifyProcessInstanceCommandImpl.java [hunk 4]
```diff
@@ -150,13 +151,34 @@ public ModifyProcessInstanceCommandStep3 withVariables(
...
+  public ModifyProcessInstanceCommandStep3 withVariable(final String key, final Object value) {
+    return withVariable(key, value, EMPTY_SCOPE_ID);
+  }
+
+  @Override
+  public ModifyProcessInstanceCommandStep3 withVariable(
+      final String key, final Object value, final String scopeId) {
+    final VariableInstruction variableInstruction = createVariableInstruction(key, value, scopeId);
+    addVariableInstructionToLatestActivateInstruction(variableInstruction);
+    return this;
```
Comment (remcowesterhoud) [inline clients/java/src/main/java/io/camunda/zeebe/client/impl/command/ModifyProcessInstanceCommandImpl.java:164]: ❌ Let's reuse the code we already have:

```java
return withVariables(Collections.singletonMap(key, value), scopeId);
```

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
Thanks @nicpuppa 🚀 

I've requested some changes, please have a look at them 👀

File: clients/java/src/main/java/io/camunda/zeebe/client/impl/command/ModifyProcessInstanceCommandImpl.java (hunk 3 above)
Comment (nicpuppa) [reply clients/java/src/main/java/io/camunda/zeebe/client/impl/command/ModifyProcessInstanceCommandImpl.java:177]: `null` is also a valid key, do you think is better to remove also this check ?

Comment (remcowesterhoud) [reply clients/java/src/main/java/io/camunda/zeebe/client/impl/command/ModifyProcessInstanceCommandImpl.java:177]: No I think we should keep this one. We shouldn't encourage people using null as a key 😄

**Review by remcowesterhoud — APPROVED** [review]
Thanks Nic! 🙇 

bors merge

**Pull request merged**